src/Spice86.Core/Emulator/InterruptHandlers/Common/MemoryWriter/MemoryAsmWriter.cs

```diff
@@ -0,0 +1,67 @@
+namespace Spice86.Core.Emulator.InterruptHandlers.Common.MemoryWriter;
+
+using Spice86.Core.Emulator.InterruptHandlers.Common.Callback;
+using Spice86.Core.Emulator.Memory;
+using Spice86.Shared.Emulator.Memory;
+
+/// <summary>
+/// Writes x86 ASM instructions to Memory bus
+/// </summary>
+public class MemoryAsmWriter : MemoryBytesWriter {
+    private CallbackHandler _callbackHandler;
+
+    public MemoryAsmWriter(Memory memory, SegmentedAddress beginningAddress, CallbackHandler callbackHandler) : base(memory, beginningAddress) {
+        _callbackHandler = callbackHandler;
+    }
+
+    /// <summary>
+    /// Registers a new callback that will call the given runnable:
+    ///  - Callback will know its physical address in memory
+    ///  - Callback will be registered in the callback handler
+    ///  - Callback instruction referring this callback will be written as ASM
+    /// </summary>
+    /// <param name="callbackNumber">Callback index</param>
+    /// <param name="runnable">Action to run when this callback is executed by the CPU</param>
+    public void RegisterAndWriteCallback(byte callbackNumber, Action runnable) {
+        uint physicalAddress = CurrentAddress.ToPhysical();
+        ICallback callback = new Callback(callbackNumber, runnable, physicalAddress);
+        _callbackHandler.AddCallback(callback);
+        WriteCallback(callback.Index);
+    }
+
+    private void WriteCallback(byte callbackNumber) {
+        WriteByte(0xFE);
+        WriteByte(0x38);
+        WriteByte(callbackNumber);
+    }
+
+    public void WriteIret() {
+        WriteByte(0xCF);
+    }
+
+    public void WriteNop() {
+        WriteByte(0x90);
+    }
+
+    /// <summary>
+    /// Writes a far CALL instruction. 
+    /// </summary>
+    /// <param name="destination">Segmented address of the function to call</param>
+    /// <returns>Returns the address of the call destination segmented address</returns>
+    public SegmentedAddress WriteFarCall(SegmentedAddress destination) {
+        WriteByte(0x9A);
+        return WriteSegmentedAddress(destination);
+    }
+
+    public SegmentedAddress WriteSegmentedAddress(SegmentedAddress address) {
+        // Make a copy of the address since it is going to be modified by our writes.
+        SegmentedAddress ret = new SegmentedAddress(CurrentAddress);
```

Comment:
If a copy is never meant to be modified, maybe we could have a ValueSegmentedAddress which is a record struct.

(if it can be mutated, then it's simply a struct with the same interface, ie. Segment, Offset, etc...)

Therefore, the copy will live on the stack, and the GC won't have to deal with it.
It will be freed when it goes out of scope.

The emulator creates a lot of SegmentedAddresses.

However, this is probably an over-optimization.

You could also copy the Segment and Offset into a value tuple:

```
(ushort Offset, ushort Segment) = (CurrentAddress.Offset, CurrentAddress.Segment)

Reply:
Would be nice for general use, we have a lot of places where we deal with segement / offset params separately to avoid an allocation. However here this code is called only a few times before execution starts so it's not going to break GC :D